### DISCUSSION TITLE
Running Pydantic path validations before a `Depends` block

### DISCUSSION THREAD

**Discussion by alvarotroya** [fastapi/fastapi#11846] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uuid

import fastapi

app = fastapi.FastAPI()


def get_db():
    raise ValueError("This is a DB error")


@app.get("/{some_id}")
async def root(some_id: uuid.UUID, _db=fastapi.Depends(get_db)):
    return {"some_id": some_id}

# Used `curl localhost:8000/not_a_uuid` to test
# Endpoint raises the ValueError in the dependency instead of a RequestValidationError due to the malformed uuid

```

### Description

Hi everyone,

tldr; Is there a way to have pydantic path validations, e.g. a uuid validation as in the example above, run __before__ FastAPI exeutes my dependency?

More context: Due to a malformed UUID being passed in the path of my endpoint(s), the `fastapi.Depends` callable is raising a cryptic error in my API logs and passing a 500 to the client. Before introducing the dependency, the usual pydantic+fastAPI awesomeness took care of returning a 422 with an 'Invalid UUID format' message to the client. 

While I can see that it can be useful to have dependencies run before a request has been "type"-validated, there are also use cases where you would rather expect that a `Depends` is only resolved __after__ the type validations. Is there a way to achieve this?

I haven't been able answer this question by searching the documentation or GitHub discussions or looking into the code. Apologies in advance if this is written up somewhere.

And of course many thanks for the help, any hint is very much appreciated!

Edit: I can see that validating the UUID in my dependency is a workaround for this example, however I am a bit worried about what could go wrong if other "bad" requests are passed unvalidated into the Depends.

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.11

**Comment by YuriiMotov** [comment]
As I know you can not change the order in which FastAPI resolves dependencies. FastAPI determines the order by creating a tree of sub-dependencies and starts resolving it from the top.

So, the only way to change the order is to move (or duplicate) the `some_id` parameter to `get_db`:

```
import uuid
import fastapi

app = fastapi.FastAPI()

def get_db(some_id: uuid.UUID):
    raise ValueError("This is a DB error")

@app.get("/{some_id}")
async def root(some_id: uuid.UUID, _db=fastapi.Depends(get_db)):
    return {"some_id": some_id}
```

As I understand what you want is to force FastAPI to resolve dependencies in two passes: 1) resolve (validate) input parameters, 2) resolve all other dependencies.

**Comment by YuriiMotov** [comment, marked as answer]
I got the idea how to solve the issue in this particular case.

Just specify the type of that path parameter in path string:

```
@app.get("/{some_id:uuid}")  # specify that `some_id` should be of type `uuid`
```

But it will only work for simple checks like `int`, `float`, `uuid`.
For more complex cases [this](https://github.com/fastapi/fastapi/discussions/11846#discussioncomment-10059856) is the only way to go.

Considering that this seems to be not something people often struggle with, I'm going to close this discussion for now.

If you have any real-life examples of when it might be critical, please provide such examples.
